wasm/src/cidr.c

```diff
@@ -0,0 +1,321 @@
+#include <stddef.h>
+#include <string.h>
+
+#include "str.h"
+#include "value.h"
+
+typedef struct {
+    uint8_t len;
+    unsigned char ip[16];
+    unsigned char mask[16];
+} ip_net;
+
+typedef unsigned char u_char;
+typedef unsigned int u_int;
+
+static int inet_pton4(const char *src, const char *end, u_char *dst);
+static int inet_pton6(const char *src, const char *end, u_char *dst);
+
+static bool parse_ip(const char *src, int n, ip_net *dst)
+{
+    for (int i = 0; i < n; i++) {
+        if (src[i] == '.')
+        {
+            dst->len = 4;
+            memset(dst->mask, 0xff, 4);
+            return inet_pton4(src, src + n, dst->ip);
+        }
+        else if (src[i] == ':')
+        {
+            dst->len = 16;
+            memset(dst->mask, 0xff, 16);
+            return inet_pton6(src, src + n, dst->ip);
+        }
+    }
+
+    return FALSE;
+}
+
+static bool parse_cidr(const char *src, size_t n, ip_net *dst)
+{
+    const char *slash = NULL;
+    for (size_t i = 0; i < n; i++)
+    {
+        if (src[i] == '/')
+        {
+            slash = &src[i];
+            break;
+        }
+    }
+
+    if (slash == NULL)
+    {
+        return FALSE;
+    }
+
+    const char *addr = src;
+    const size_t len = slash - src;
+    if (!parse_ip(addr, len, dst))
+    {
+        return FALSE;
+    }
+
+    const char *mask = slash + 1;
+    long long bits;
+    if (opa_atoi64(mask, n - len - 1, &bits) == -1 || bits < 0 || bits > dst->len*8)
+    {
+        return FALSE;
+    }
+
+    for (int i = 0; i < dst->len; i++)
+    {
+        if (bits >= 8) {
+            dst->mask[i] = 0xff;
+            bits -= 8;
+            continue;
+        }
+
+        dst->mask[i] = ~((unsigned char)(0xff >> bits));
+        bits = 0;
+    }
+
+    for (int i = 0; i < dst->len; i++)
+    {
+        dst->ip[i] &= dst->mask[i];
+    }
+
+    return TRUE;
+}
+
+// returns true if a contains b.
+static bool contains(ip_net *a, ip_net *b)
+{
+    if (a->len != b->len)
+    {
+        return FALSE;
+    }
+
+    for (int i = 0; i < a->len; i++)
+    {
+        if (a->mask[i] & ~b->mask[i])
+        {
+            // b mask is shorter.
+            return FALSE;
+        }
```

Comment:
Is this just an optimization? I didn't work through the bit arithmetic to convince myself it was correct.

Reply:
Added a comment.